### ISSUE TITLE
item.hitTest with an item that has no fill (transparent)

### ISSUE THREAD

**Issue by noahbass** [paperjs/paper.js#1157]
I'm looking to hit test an item that does not have a fill.

[Take this example](http://sketch.paperjs.org/#S/bZFNa8MwDIb/ivClCZg0O+SSstPOZaPb2GHdwfXUJMSxg610h5L/Ptml2Qc1+Ot5JflFPgurBhS1eO6RdCuk0O4z3k/Kg0dNcA8Wv+BJUVvs+K5sYzBLyHWWsqqUUJW5hB90VzLjJc83e7+3sUoRyLseH5xxniuuDmbCVVLjXK/hrUWLJ/RALcIU+DB6DAFDAoNjVMfQ42Q1dc6Cs9sIX8eM0yzlcI4y8IjO244exxgX+LVFiSNgM3B8qIH8hPKPlDzeEI6dMTcwOYNeWc0p1ZXPm382dhgmE7uY2sDgBQNdPBdjbJf8ZTZfsq+7ZsrPFMY12VLtEjbzZx08qj6VCaJ+/5i/AQ==) of an item that does not have a fill color. The hit test is not triggered when you click inside the item. This is expected because there is no fill color, but what if I want to do a hit test on an item with no fill color?

The only way I see to get around this is to forge a transparent color (such as `Color(0, 0, 0, 0.0001)`) with an extremely transparent rgba color [as in this case](http://sketch.paperjs.org/#S/bZFNb4MwDIb/ipVLiYQoPXCh2mnnaVO3aYd1hzRzARESlITuUPHfZ9NBt6kWH8nz2i+OOQurOhSleG4x6lqkQrtP3p+UB486wh1Y/IInFetsR3tlK4PJhFxjY1LkKRS5TOGKNjkxeki53fu9ZZcsRO9avHfGeXJcHcyAq6t6bIyZNfaZ1gm5/FxZzrGRXMD3eg1vNVo8oYdYIwyBFr3HEDBMoHOESk49DlbHxllw9oHha59QmY0SziwDBR+1buJjz3mBWlgUjoBVR/mhhOgHTP9I06FuCHyeGzg6g15ZTSXFzMftvzZ2GAbDY58mQ+AFQ7z0nPU83/RXs3Kpnt+aKH0mM65KFrdL2kh/9+BRtZNNEOX7x/gN). Now clicking inside the item triggers the hit test. Now if the color is modified to `Color(0, 0, 0, 0)`, the hit test will not trigger.

The solution seems to be to set the fill color to a low alpha such as `Color(0, 0, 0, 0, 0.0001)` where it would not be noticeable to a user. Although it works, this feels hacky.

So the question: is there a way to have an item that has a 100% transparent fill color and can be used in a hit test?

**Comment by radpet** [comment]
Paper.js triggered hitTest when the shape was with no fill color in v0.9.25

**Comment by prichey** [comment]
I'd like to keep up the discussion from https://github.com/paperjs/paper.js/issues/1218.

Items with a transparent fill (any color with an opacity of 0) are considered not to be filled. See here: https://github.com/paperjs/paper.js/blob/develop/src/style/Style.js#L312. (This applies to stroke and shadow as well.) Therefore, hitTest doesn't return a hit on an item's fill (if it is transparent), because it is considered not to have a fill in the first place.

I'm wondering if this is the correct decision. In my mind, an item with a transparent fill should be distinguished from an item without a fill at all, though that could be more influenced by raster thinking than vector. Anybody else have thoughts?

**Comment by simonemazzoni** [comment]
@prichey I agree with you. I'm facing problems right now since I can't hitTest a shape using its fill if it was created without it.

I found also the function `isInside(rect)` in the `Point` class, that does kid of the the check I need to verify if a point is inside a shape, except for the fact that I check it against The shape bounds and not against the actual shape.
This is not ideal and force me to write additional code together with the usual hitTest, so I think that the hitTest code need a refactor to include also the shape fill even if it is transparent.

Did you find a better way other than setting an extremely transparent fill? @noahbass

**Comment by m5x** [comment]
We too have the same problem. We have created a simple vector editor using paper.js. Up until now we've been using version 0.9.25 so the hit test worked fine for our use case but now after we upgraded to latest version there seems to be no way to hit test shapes without a fill or with a transparent fill. Simple solution of using color with alpha = 0.001 or something like that comes to mind but it'd be great if paper allowed to optionally hit test fill of shapes without a fill.

**Comment by sasensi** [comment]
A less hacky workaround is possible using [project.getItems({match:function(){}})](http://paperjs.org/reference/project/#getitems-options) to filter project items based on wether they contain event point or not with [item.contains()](http://paperjs.org/reference/item/#contains-point) method. This is actually was is used in hit test code: https://github.com/paperjs/paper.js/blob/b1705f628b829fe296d8191adbb8016fea09dc49/src/item/Item.js#L2095-L2099
Here is a [sketch](http://sketch.paperjs.org/#V/0.11.8/S/bVNNb5wwEP0rIy6w6pbspVJF1dP20kOkSO0tycE1A+sANrKHpVLEf++M+YxUSwh53vj5zePxnljVYVIkvxokfUvOiXal7B8eoPRqBAXaeN3ii3+xFkd4UnTLr7GUvUsReGm0hL6Au8ExnzfnFfOqNEMo4MtlKwXyrsGrax2fSZ1XtsZ0QyvTtitGjIVeeaZMBZ9O3+QlDwt0Fjo3BITSjTbPcylXg9VkGHH2UbAfDGV4Z4ITbHr5LLdXzoNqW+i9e0NNYAi7sNBI1135WIPva0teI/2Urn30na5B7IFujvVoZ0kZa2wN8WronbF0oJbVKTa8gFVwJlcdNK7LIw3eRiH5whvmgfLIOjuydk8fR1S2FAJv+ABrQ2hVILYGIZNd5zyKi+T603pwOjFvn+1GL2ymAjXLgFEFuJnjPAz+T/8sovcsk4PUYQiqRjBWDApuDtXau5Ty1tVZerwm3SeM002HBMSIsiHkh2hi2GIq1vzGv3QIKZvHrnGqrq3RjRwz5ezKIWZL3EGcC3gUy52KRBCYACUS5wHLPbbxY3z4BeATPF/O8Pnr5XXrehsCmcpoJWpZytx5zDb/gn88qibyhaR4fp3+AQ==) demonstrating the solution.
The core part is:
```javascript
// ...for all project items...
var item = project.getItems({
    // ...keep those containing event point...
    match: function(item) {
        return item.contains(event.point);
    }
// ...and retrieve the last one (the more on top)
}).pop();
```
